[add tencentcloud_sqlserver_db resource]

brickzzhang opened the pull request:
1. add tencentcloud_sqlserver_db resource
2. example contents depend on sqlserver instance

[commiter: brickzzhang]

-------

➜  terraform-provider-tencentcloud git:(master) make testacc TESTARGS=" -run=TestAccTencentCloudSqlserverDB_basic_and_update"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccTencentCloudSqlserverDB_basic_and_update -timeout 120m
?   	github.com/terraform-providers/terraform-provider-tencentcloud	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-tencentcloud/gendoc	(cached) [no tests to run]
=== RUN   TestAccTencentCloudSqlserverDB_basic_and_update
--- PASS: TestAccTencentCloudSqlserverDB_basic_and_update (241.94s)
PASS
ok  	github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud	242.744s
?   	github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/connectivity	[no test files]
?   	github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/internal/helper	[no test files]
?   	github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/ratelimit	[no test files]

Commit:
[add tencentcloud_sqlserver_db resource]

1. add tencentcloud_sqlserver_db resource
2. example contents depend on sqlserver instance

[commiter: brickzzhang]

gailwang commented on examples/tencentcloud-sqlserver/main.tf:
```diff
@@ -1,2 +1,9 @@
 }
+
+resource "tencentcloud_sqlserver_db" "mysqlserver_db" {
+  instance_id = "mssql-3cdq7kx5"
```
写example的时候用tf里创建的example instance，否则客户拿到跑不起来

gailwang commented on tencentcloud/resource_tc_sqlserver_db.go:
```diff
@@ -0,0 +1,170 @@
+	if err := sqlserverService.CreateSqlserverDB(ctx, instanceID, dbName, charset, remark); err != nil {
+		return err
+	}
+
```
create里没有写SetId吗

gailwang commented on tencentcloud/resource_tc_sqlserver_db.go:
```diff
@@ -0,0 +1,170 @@
+	_ = d.Set("create_time", *dbInfo.CreateTime)
+	_ = d.Set("status", SQLSERVER_DB_STATUS[*dbInfo.Status])
+
+	d.SetId(name)
```
为什么在Read里SetID？

gailwang commented on tencentcloud/resource_tc_sqlserver_db.go:
```diff
@@ -0,0 +1,170 @@
+
+	sqlserverService := SqlserverService{client: meta.(*TencentCloudClient).apiV3Conn}
+	instanceId := d.Get("instance_id").(string)
+
```
如果是分开设置更新，需要用到
d.Partial(true)
//update1 ...
//update2 ...
d.Partial(false)

gailwang commented on tencentcloud/resource_tc_sqlserver_db.go:
```diff
@@ -0,0 +1,170 @@
+	if d.HasChange("name") {
+		oldValue, newValue := d.GetChange("name")
+		if err := sqlserverService.ModifySqlserverDBName(ctx, instanceId, oldValue.(string), newValue.(string)); err != nil {
+			return err
```
如果name是可以修改的，为什么要拿他作为唯一ID？

gailwang commented on tencentcloud/resource_tc_sqlserver_db.go:
```diff
@@ -0,0 +1,170 @@
+	sqlserverService := SqlserverService{client: meta.(*TencentCloudClient).apiV3Conn}
+	instanceId := d.Get("instance_id").(string)
+	name := d.Get("name").(string)
+
```
有没有测试过，如果一个Instance不存在的情况下，去调用删除它的DB，会报什么错？会造成程序崩溃吗？

gailwang commented on tencentcloud/service_tencentcloud_sqlserver.go:
```diff
@@ -78,3 +82,313 @@ func (me *SqlserverService) DescribeProductConfig(ctx context.Context, zone stri
+	dbRemark.Name = &dbname
+	dbRemark.Remark = &remark
+	dbRemarkList = append(dbRemarkList, &dbRemark)
+	request.DBRemarks = dbRemarkList
```
request.DBRemarks = []*sqlserver.DBRemark{{Name:&dbName, Remark:&remark}}

gailwang commented on tencentcloud/service_tencentcloud_sqlserver.go:
```diff
@@ -78,3 +82,313 @@ func (me *SqlserverService) DescribeProductConfig(ctx context.Context, zone stri
+	}()
+
+	var response *sqlserver.DeleteDBResponse
+	err := resource.Retry(10*writeRetryTimeout, func() *resource.RetryError {
```
这里的10是测试过一定要用这么长的等待时间吗

gailwang commented on tencentcloud/service_tencentcloud_sqlserver.go:
```diff
@@ -78,3 +82,313 @@ func (me *SqlserverService) DescribeProductConfig(ctx context.Context, zone stri
+		ratelimit.Check(request.GetAction())
+		result, e := me.client.UseSqlserverClient().DeleteDB(request)
+		if e != nil {
+			log.Printf("[CRITAL]%s DescribeDBs fail, reason:%s\n", logId, e.Error())
```
Describe这里是不是有问题，这里的action用%s , request.GetAction 就不用来回改了

gailwang commented on tencentcloud/resource_tc_sqlserver_db.go:
```diff
@@ -0,0 +1,170 @@
+		d.SetId("")
+		return nil
+	}
+	_ = d.Set("charset", *dbInfo.Charset)
```
这里不用*

gailwang commented on tencentcloud/resource_tc_sqlserver_db.go:
```diff
@@ -0,0 +1,170 @@
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Database status. Valid values are `creating`, `running`, `modifying`, `dropping`.",
+			},
```
Valid value的意思是要输入的时候输入这些，现在这个参数是status，告诉客户分别代表什么含义

gailwang commented on tencentcloud/resource_tc_sqlserver_db.go:
```diff
@@ -0,0 +1,170 @@
+			"name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Name of DB. The DataBase name must be unique and must be composed of numbers, letters and underlines, and the first one can not be underline.",
```
Name of the SQL Server DB

Commit:
[enhancement for sqlserver db resource]

1. add import func
2. modify id generation way
3. modify description
4. modify test func

brickzzhang commented:
➜  terraform-provider-tencentcloud git:(master) ✗ make testacc TESTARGS=" -run=TestAccTencentCloudSqlserverDB_basic_and_update"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccTencentCloudSqlserverDB_basic_and_update -timeout 120m
?     github.com/terraform-providers/terraform-provider-tencentcloud  [no test files]
testing: warning: no tests to run
PASS
ok    github.com/terraform-providers/terraform-provider-tencentcloud/gendoc (cached) [no tests to run]
=== RUN   TestAccTencentCloudSqlserverDB_basic_and_update
--- PASS: TestAccTencentCloudSqlserverDB_basic_and_update (193.43s)
PASS
ok    github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud 194.689s
?     github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/connectivity  [no test files]
?     github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/internal/helper [no test files]
?     github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/ratelimit [no test files]

gailwang committed:
Merge branch 'master' into master
